src/plugins/csvstorage/csvstorage.c

```diff
@@ -287,34 +300,66 @@ int elektraCsvstorageGet(Plugin *handle, KeySet *returned, Key *parentKey)
 		const char *delimString = keyString(delimKey);
 		delim = delimString[0];
 	}
-	
+
 	Key *readHeaderKey = ksLookupByName(config, "/useheader", 0);
 	short useHeader = 0;
 	if(readHeaderKey)
 	{
 		const char *printHeaderString = keyString(readHeaderKey);
-		if(strcmp(printHeaderString, "0"))
+		if(!strcmp(printHeaderString, "1"))
 		{
 			useHeader = 1;
 		}
+		else if(!(strcmp(printHeaderString, "-1")))
+		{
+			useHeader = -1;
+		}
+		else
+			useHeader = 0;
 	}
 	unsigned long fixColumnCount = 0;
 	Key *fixColumnCountKey = ksLookupByName(config, "/columns", 0);
 	if(fixColumnCountKey)
 	{
-	    if(keyString(fixColumnCountKey))
-	    {
-		fixColumnCount = atol(keyString(fixColumnCountKey));
-	    }
+		if(keyString(fixColumnCountKey))
+		{
+			fixColumnCount = atol(keyString(fixColumnCountKey));
+		}
+	}
+	Key *setNamesKey = ksLookupByName(config, "/colNames", 0);
+	char *colNames = NULL;
+	if(setNamesKey)
+	{
+		if(fixColumnCountKey)
+		{
+			KeySet *namesKS = ksCut(config, setNamesKey);
+			ssize_t nrNames = ksGetSize(namesKS)-1;
+			if(nrNames == fixColumnCount)
+			{
+				colNames = elektraMalloc(nrNames*sizeof(char *));
+				Key *cur;
+				char **ptr = colNames;
+				while((cur = ksNext(namesKS)) != NULL)
+				{
+					if(!strcmp(keyName(cur), keyName(setNamesKey))) continue;
+					if(!strcmp(keyString(cur), ""))
+						*ptr = NULL;
+					else
+						*ptr = keyString(cur);
+					++ptr;
+				}	
+			}
+		}
 	}
 	int nr_keys;
-	nr_keys = csvRead(returned, parentKey, delim, useHeader, fixColumnCount);
-
+	nr_keys = csvRead(handle, returned, parentKey, delim, useHeader, fixColumnCount, colNames);
+	if(colNames)
```

Comment:
not sure where you mean. i could free it inside csvRead, but i think it's easier to read / find if it's within the same function it's allocated.